lib/ipmi_delloem.c

```diff
@@ -338,6 +365,689 @@ usage(void)
+	printf("Device Revision     : %0.2d\n", dev_id->device_revision & 0x7F);
+	if( (iDRAC_FLAG == IDRAC_12G) || (iDRAC_FLAG == IDRAC_13G) || (iDRAC_FLAG == IDRAC_14G)) {
+		if(firmware_ver.major_ver == 0) {
+			printf("Firmware Version    : %d.%0.2d(Build %02d)\n", dev_id->fw_rev1 & 0x7F,
+				dev_id->fw_rev2,dev_id->aux_fw_rev[1]);
+		} else {
+			if((iDRAC_FLAG == IDRAC_13G) || (iDRAC_FLAG == IDRAC_14G)) {
+				printf("Firmware Version    : %d.%0.2d.%02d.%02d(Build %02d)\n", firmware_ver.major_ver,
+				firmware_ver.minor_ver,(firmware_ver.pla_data_msb << 8 | firmware_ver.pla_data_lsb),
+				(firmware_ver.lc_ver_msb << 8 | firmware_ver.lc_ver_lsb),
+				(firmware_ver.build_no_msb << 8 | firmware_ver.build_no_lsb));
+			} else {
+				printf("Firmware Version    : %d.%0.2d.%02d(Build %02d)\n", firmware_ver.major_ver,
+				firmware_ver.minor_ver,(firmware_ver.pla_data_msb << 8 | firmware_ver.pla_data_lsb),
+				(firmware_ver.build_no_msb << 8 | firmware_ver.build_no_lsb));
+			}
+		}
+	}else {
+		// DF427387 Fix for Build Version Number
+		//	Byte:	  13				14		   15	       16
+		//	Data: <platform ID> <Build Number> <RES> <RES>
+		printf("Firmware Version    : %d.%0.2x.%02d\n", dev_id->fw_rev1 & 0x7F,
+			dev_id->fw_rev2,dev_id->aux_fw_rev[1]);
+	}
+
+	/* BCD value. byte->hex conversion used.*/
+	printf("IPMI Version        : %d.%d\n", dev_id->ipmi_version & 0x0F, dev_id->ipmi_version >> 4);
+	ipmi_version = dev_id->ipmi_version;
+
+
+	printf("Manufacturer ID     : %lu\n",
+		(uint32_t)(dev_id->manufacturer_id[2] & 0x0f) * 65536 +
+		(uint32_t)(dev_id->manufacturer_id[1] * 256) +
+		(uint32_t)dev_id->manufacturer_id[0]);
+	printf("Product ID          : %d\n", dev_id->product_id[1]*256 + dev_id->product_id[0]);
+
+	memset(&req, 0, sizeof(req));
+	req.msg.netfn = IPMI_NETFN_APP;
+	req.msg.lun = 0;
+	req.msg.cmd = BMC_GET_SELF_TEST;
+	req.msg.data = NULL;
+	req.msg.data_len = 0;
+	rsp = intf->sendrecv(intf, &req);
+
+	if (rsp->data[0] == 0x55) {
+	printf("Status              : No error\n",-1);
+	}else if (rsp->data[0] == 0x56) {
+	printf("Status              : Self Test Function not implemented in this controller\n");
+	}else if (rsp->data[0] == 0x58)
+	{ printf("Status              : Fatal Hardware Error (BMC is inoperative)\n");
+	}else if (rsp->data[0] == 0x57) {
+		printf("Status              : ");
+		if (rsp->data[1] & 0x01) {printf("Controller operational firmware corrupte "); }
+		if (rsp->data[1] & 0x02) {printf("Controller update 'boot block' firmware corrupte "); }
+		if (rsp->data[1] & 0x04) {printf("Internal Use Area of BMC FRU corrupted "); }
+		if (rsp->data[1] & 0x08) {printf("SDR Repository Empty "); }
+		if (rsp->data[1] & 0x10) {printf("IPMB signal lines do not respond "); }
+		if (rsp->data[1] & 0x20) {printf("Cannot access BMC FRU device "); }
+		if (rsp->data[1] & 0x40) {printf("Cannot access SDR Repository "); }
+		if (rsp->data[1] & 0x80) {printf("Cannot access SEL device"); }
```

Comment:
No. Magic numbers are not acceptable. No `0x56`, `0x40` or whatever. Descriptive macros please.